syz-verifier/rpcserver.go

```diff
@@ -86,127 +81,74 @@ func (srv *RPCServer) UpdateUnsupported(a *rpctype.UpdateUnsupportedArgs, r *int
 		vrf.SetPrintStatAtSIGINT()
 
 		vrf.choiceTable = vrf.target.BuildChoiceTable(nil, vrf.calls)
-		srv.cond.Signal()
+		vrf.progGeneratorInit.Done()
+
 	}
 	return nil
 }
 
 // NextExchange is called when a Runner requests a new program to execute and,
 // potentially, wants to send a new Result to the RPCServer.
 func (srv *RPCServer) NextExchange(a *rpctype.NextExchangeArgs, r *rpctype.NextExchangeRes) error {
-	srv.mu.Lock()
-	defer srv.mu.Unlock()
-
-	var res *ExecResult
-	var prog *progInfo
 	if a.Info.Calls != nil {
-		res = &ExecResult{
-			Pool:   a.Pool,
-			Hanged: a.Hanged,
-			Info:   a.Info,
-			RunIdx: a.RunIdx,
-		}
-
-		prog = srv.progs[a.ProgIdx]
-		if prog == nil {
-			// This case can happen if both conditions are true:
-			// 1. a Runner calls Verifier.NextExchange, then crashes,
-			// its corresponding Pool being the only one that hasn't
-			// sent results for the program yet
-			// 2.the cleanup call for the crash got the server's mutex before
-			// the NextExchange call.
-			// As the pool was the only one that hasn't sent the result, the
-			// cleanup call has already removed prog from srv.progs by the time
-			// the NextExchange call gets the server's mutex, which is why the
-			// variable is nil. As the results for this program have already
-			// been sent for verification, we discard this one.
-			return nil
-		}
-
-		delete(srv.pools[a.Pool].runners[a.VM], prog.idx)
-		if srv.newResult(res, prog) {
-			if srv.vrf.processResults(prog) {
-				delete(srv.progs, prog.idx)
-			}
-		}
+		srv.stopWaitResult(a.Pool, a.VM, a.ExecTaskID)
+		PutExecResult(&ExecResult{
+			Pool:       a.Pool,
+			Hanged:     a.Hanged,
+			Info:       a.Info,
+			ExecTaskID: a.ExecTaskID,
+		})
 	}
 
-	if srv.notChecked > 0 {
-		// Runner is blocked until the choice table is created.
-		srv.cond.Wait()
+	// TODO: NewEnvironment is the currently hardcoded logic. Relax it.
+	r.Id = rpctype.NoTask
+	task := srv.vrf.GetRunnerTask(a.Pool, NewEnvironment)
+	if task != nil {
```

Comment:
GetRunnerTask does not seem to ever return nil. Does it?

Reply:
No. Cleanup done.

Reply:
Not now. Cleaned code here.